### ISSUE TITLE
Incorrect and inconsistent WCS transformations with both sliced WCS and Cutout2D when SIP is present

### ISSUE THREAD

**Issue by mcara** [astropy/astropy#7550]
I've been trying to get a WCS slice through several methods (slicing and `Cutout2D`) but I get confusing, inconsistent results that do not really make sense. Let me first provide a data set so that the issue could be reproduced and then describe  steps to reproduce the results.

### Content of wcshdr.txt:

```
NAXIS   =                    2                                                  
NAXIS1  =                 1014                                                  
NAXIS2  =                 1014                                                  
WCSAXES =                    2 / Number of coordinate axes                      
CRPIX1  =                507.0 / Pixel coordinate of reference point            
CRPIX2  =                507.0 / Pixel coordinate of reference point            
PC1_1   =    7.70701846115E-06 / Coordinate transformation matrix element       
PC1_2   =    3.29123708066E-05 / Coordinate transformation matrix element       
PC2_1   =     3.6822687421E-05 / Coordinate transformation matrix element       
PC2_2   =   -6.77397792858E-06 / Coordinate transformation matrix element       
CDELT1  =                  1.0 / [deg] Coordinate increment at reference point  
CDELT2  =                  1.0 / [deg] Coordinate increment at reference point  
CUNIT1  = 'deg'                / Units of coordinate increment and value        
CUNIT2  = 'deg'                / Units of coordinate increment and value        
CTYPE1  = 'RA---TAN-SIP'       / TAN (gnomonic) projection + SIP distortions    
CTYPE2  = 'DEC--TAN-SIP'       / TAN (gnomonic) projection + SIP distortions    
CRVAL1  =        251.204256787 / [deg] Coordinate value at reference point      
CRVAL2  =        57.5818199439 / [deg] Coordinate value at reference point      
LONPOLE =                180.0 / [deg] Native longitude of celestial pole       
LATPOLE =        57.5818199439 / [deg] Native latitude of celestial pole        
WCSNAME = 'T1'                 / Coordinate system title                        
CRDER1  =    9.65237944683E-07 / [deg] Random error in coordinate               
CRDER2  =    6.27414003418E-07 / [deg] Random error in coordinate               
RADESYS = 'ICRS'               / Equatorial coordinate system                   
A_ORDER =                    4                                                  
A_0_2   = 5.33092691853019E-08                                                  
A_0_3   = 3.73753772880087E-11                                                  
A_0_4   = -2.0211147283637E-13                                                  
A_1_1   = 2.44084308174762E-05                                                  
A_1_2   = 2.81394789152228E-11                                                  
A_1_3   = 5.17856894659408E-13                                                  
A_2_0   = -2.4133465663211E-07                                                  
A_2_1   = 1.29289254704036E-10                                                  
A_2_2   = 2.35753629384072E-14                                                  
A_3_0   =  -2.371620072818E-10                                                  
A_3_1   = 5.43714947358335E-13                                                  
A_4_0   = -2.8102976693048E-13                                                  
B_ORDER =                    4                                                  
B_0_2   = 2.99270373714981E-05                                                  
B_0_3   = -2.3813607426051E-10                                                  
B_0_4   = 7.23205168173609E-13                                                  
B_1_1   = -1.7107385824843E-07                                                  
B_1_2   = 6.31243430776636E-11                                                  
B_1_3   = -5.1674434652779E-14                                                  
B_2_0   = 6.95458963321812E-06                                                  
B_2_1   = -3.0827896112996E-10                                                  
B_2_2   = -1.7580091715687E-13                                                  
B_3_0   = 3.51974158902385E-11                                                  
B_3_1   = 5.60993015610249E-14                                                  
B_4_0   = -5.9243852454034E-13                                                  
```

### 1. Load the header and create a "full" (as opposite to a cutout) WCS:

```python
>>> from astropy.io import fits
>>> from astropy import wcs
>>> wcshdr = fits.Header.fromtextfile('wcshdr.txt')
>>> fullwcs = wcs.WCS(wcshdr)
>>> print(fullwcs)
WCS Keywords

Number of WCS axes: 2
CTYPE : 'RA---TAN-SIP'  'DEC--TAN-SIP'  
CRVAL : 251.20425678699999  57.581819943900001  
CRPIX : 507.0  507.0  
PC1_1 PC1_2  : 7.7070184611500002e-06  3.2912370806599997e-05  
PC2_1 PC2_2  : 3.6822687421e-05  -6.7739779285800001e-06  
CDELT : 1.0  1.0  
NAXIS : 1014  1014
```

### 2. Make a "slice" around `CRPIX` of size `32x32`

```python
>>> slwcs = fullwcs[491:523, 491:523]
>>> print(slwcs)
WCS Keywords

Number of WCS axes: 2
CTYPE : 'RA---TAN-SIP'  'DEC--TAN-SIP'  
CRVAL : 251.20425678699999  57.581819943900001  
CRPIX : 16.0  16.0  
PC1_1 PC1_2  : 7.7070184611500002e-06  3.2912370806599997e-05  
PC2_1 PC2_2  : 3.6822687421e-05  -6.7739779285800001e-06  
CDELT : 1.0  1.0  
NAXIS : 32  32
```

### 3. Make a "similar" slice using `Cutout2D`:

```python
>>> from astropy.nddata import Cutout2D
>>> cutout = Cutout2D(np.ones(fullwcs._naxis), (491, 491), (32, 32), fullwcs.deepcopy())
>>> ctwcs = cutout.wcs
>>> print(ctwcs)
WCS Keywords

Number of WCS axes: 2
CTYPE : 'RA---TAN-SIP'  'DEC--TAN-SIP'  
CRVAL : 251.20425678699999  57.581819943900001  
CRPIX : 32.0  32.0  
PC1_1 PC1_2  : 7.7070184611500002e-06  3.2912370806599997e-05  
PC2_1 PC2_2  : 3.6822687421e-05  -6.7739779285800001e-06  
CDELT : 1.0  1.0  
NAXIS : 1014  1014
```

Notice that `ctwcs` has different `CRPIX` and `NAXIS` (it does not adjust `_naxis` attribute) compared to the WCS obtained through slicing (`slwcs`).

### 4. Check that `CRVAL == wcs.all_pix2world(CRPIX)`

```python
>>> # check for the WCS obtained through slicing:
>>> slwcs.all_pix2world(*slwcs.wcs.crpix, 1) - slwcs.wcs.crval
array([ 0.00063097,  0.00015619])
>>> # check for the WCS obtained through Cutout2D:
>>> ctwcs.all_pix2world(*ctwcs.wcs.crpix, 1) - ctwcs.wcs.crval
array([ 0.0005904,  0.0001461])
```

These are some very large errors (IMO) compared to pixel scale:
```python
>>> wcs.utils.proj_plane_pixel_scales(fullwcs)
array([  3.76205854e-05,   3.36022459e-05])
```

On top of that, these world coordinates (well, deltas relative to `CRVAL`) are different for the two sliced/cutout WCS!

### 5. Check that `CRVAL == wcs.wcs_pix2world(CRPIX)` (ignoring SIP):

```python
>>> ctwcs.wcs_pix2world(*ctwcs.wcs.crpix, 1) - ctwcs.wcs.crval
array([ 0.,  0.])
>>> slwcs.wcs_pix2world(*slwcs.wcs.crpix, 1) - slwcs.wcs.crval
array([ 0.,  0.])
```
... as expected.

### Conclusions:

1. Let's start with the easy part: WCS objects obtained through slicing and `Cutout2D` have inconsistent `CRPIX` and `NAXIS` values. Not adjusting `NAXIS` in `Cutout2D` is likely an omission but difference in `CRPIX` could be more serious.

2. Transformations from `CRPIX` of the sliced/cutout WCS to world coordinates using full (including SIP) transformations does not result in `CRVAL` of these sliced or cutout WCS.

3. Transformations from `CRPIX` of the sliced/cutout WCS to world coordinates **ignoring SIP**  corrections do recover the `CRVAL` of these sliced or cutout WCS.

This last point suggests to me that sliced/cutout WCS do not properly handle SIP: at `CRPIX` distortions should vanish and so the results from both `all_pix2world()` and `wcs_pix2world()` should be identical. To verify that `all_pix2world()` works correctly for the "full" WCS, I performed one last test:

```python
>>> fullwcs.all_pix2world(*fullwcs.wcs.crpix, 1) - fullwcs.wcs.crval
array([ 0.,  0.])
```

CC: @keflavich @larrybradley @astrofrog

**Comment by pllim** [comment]
Looks like a bug *SIP*ped through. :smirk:

**Comment by larrybradley** [comment]
@mcara `Cutout2D` makes a cutout from a *central* position.  To directly compare to WCS slice in your example, you need to use this:  (also note that there is a typo in line 4 in your example -- it should be
``fullwcs = wcs.WCS(hdrwcs)``):

```
cutout = Cutout2D(np.ones(fullwcs._naxis), (507, 507), (32, 32), fullwcs.copy())
print(cutout.wcs)
WCS Keywords

Number of WCS axes: 2
CTYPE : 'RA---TAN-SIP'  'DEC--TAN-SIP'  
CRVAL : 251.204256787  57.5818199439  
CRPIX : 16.0  16.0  
PC1_1 PC1_2  : 7.70701846115e-06  3.29123708066e-05  
PC2_1 PC2_2  : 3.6822687421e-05  -6.77397792858e-06  
CDELT : 1.0  1.0  
NAXIS : 1014  1014
```

That explains the *inconsistency* part -- they are actually consistent.  But, they both may be incorrect. :)

When I had discussed WCS cutouts with you before (long ago), I think we assumed that SIP applies the polynomial corrections to `(pixel - CRPIX)` (i.e. the pixel *deltas*) so the thought was that shifting the CRPIX for the cutout should work with SIP.  Perhaps that's not right.

**Comment by larrybradley** [comment]
I fixed the `Cutout2D` `NAXIS` in #7552.

**Comment by mcara** [comment]
@larrybradley (About `pix-CRPIX`) I just conveyed what I read, e.g., in https://fits.gsfc.nasa.gov/registry/sip/shupeADASS.pdf - second paragraph in second section. I also remember when, at one point, I waded into the code, it kind of confirmed what I read in the paper. At the very least it did not rise any alarms.

**Comment by mcara** [comment]
@larrybradley I am also happy that slicing is at least consistent (and thanks for clarifying the usage of `Cutout2D`).

**Comment by mcara** [comment]
I actually wonder if this is an issue with slicing or it is some bug/change in the core `astropy.wcs`. Could someone test this by repeating my tests using some older (1yr, 2yrs, 3yrs back) versions of `astropy`. It could be that something has changed internally in the `WCS` that older assumptions are no longer valid...

**Comment by pllim** [comment]
@mcara , feel free to install version as old as you like via `pip` (see https://pypi.org/project/astropy/#history) and see for yourself. :wink:

**Comment by mcara** [comment]
@pllim Thanks for pointers! I tested this with `astropy 1.1` and I got the same behavior.

**Comment by mcara** [comment]
So, I figured what is going on here. If you look at the last parameter in the [`Sip` class](http://docs.astropy.org/en/stable/api/astropy.wcs.Sip.html), you will notice that it has its own `crpix` parameter. Essentially, `Sip.crpix` is independent from `WCS.wcs.crpix`. This is not in the standard (or, required by the standard): `Sip.crpix` is never written to the header. Instead this is a feature of `astropy.wcs.Sip`. So, when `WCS.wcs.crpix` the two `crpix`es get out of sync. Therefore, when you adjust `WCS.wcs.crpix`, you should also adjust `Sip.crpix`.

Without this synchronization, SIP corrections are computed relative to the old `CRPIX` and depending on the position of the cutout/slice relative to original `CRPIX`, this can lead to very large errors: tens of pixels for HST WFC/HRC image.

To check this, do the following (after running commands from https://github.com/astropy/astropy/issues/7550#issue-330568337):

```python
>>> print(slwcs.sip.crpix)
[ 507.  507.]
```

**Comment by larrybradley** [comment]
Fixed in #7552 and #7556.
